### ISSUE TITLE
Open accessibility questions

### ISSUE THREAD

**Issue by LukasHirt** [owncloud/web#3892]
- [x] Accessibility of infinite scroll tables (files list) #2426
- [x] Focus in the modal - how to make people happy by not having to do one extra "tab" but not cut the content from screen readers
- [x] Visible focus on elements which get focus only programmatically?

**Comment by LukasHirt** [comment]
- [ ] Running accessibility tests in the CI (is there a way to handle authorisation with axe-core?)

**Comment by michaelstingl** [comment]
> * [ ]  Running accessibility tests in the CI (is there a way to handle authorisation with axe-core?) #3495

How does other automated testing work? Nothing that can be applied here?

**Comment by LukasHirt** [comment]
> How does other automated testing work? Nothing that can be applied here?

There's an approach that the axe is included directly in the code when some new UI element appears. Not sure though how this could be used in CI because it can e.g. output the error in browser console. Also, it seems like a lot of extra code if we'd have it like that.

Looking at different tools it seems that they work pretty much in the same way.

**Comment by michaelstingl** [comment]
Maybe start with a simple tool. Better than nothing. Can be extended later…

**Comment by LukasHirt** [comment]
- [ ] Error message under input
- [ ] Password dialogue when opening public links - should focus be sent to the wrapper or to the input

**Comment by marcus-herrmann** [comment]
> [ ] Error message under input
> [ ] Password dialogue when opening public links - should focus be sent to the wrapper or to the input

To detect an error message and current focus sounds like a job for E2E testing rather than axe alone.

I have no experience with this combination, but what about https://www.npmjs.com/package/cypress-axe ?

**Comment by marcus-herrmann** [comment]
On a unit level, Jest and axe also work together:

```
it('should demonstrate this matcher`s usage', async () => {
  const render = () => '<img src="#"/>'

  // pass anything that outputs html to axe
  const html = render()

  expect(await axe(html)).toHaveNoViolations()
})
```

And checking focus: 

```
expect(document.activeElement).toBe(container.querySelector('#elementThatShouldReceiveProgrammaticFocus'))
```

**Comment by LukasHirt** [comment]
> Accessibility of infinite scroll tables (files list) #2426

- Check if the `update` event could be used
	- Might be too noisy
	- Try to log the event
- Maybe provide an option to switch to pagination
- Provide separators by alphabet and use them as skip links

> Focus in the modal - how to make people happy by not having to do one extra "tab" but not cut the content from screen readers

- Needs research if it's okay to autofocus input
	- In this use case might be a valid approach
- Push more info into the `aria-label` of the input
	- Alternative can be `aria-describedby` https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-describedby_attribute
		- Points to a preexisting id

> Visible focus on elements which get focus only programmatically?

- Look into different triggers of focus (keyboard, touch, mouse click)
- Better to keep it (with possible restriction to focus visible)

> Error message under input

- use `aria-invalid` (for errors)
- use `aria-live` (for warnings)

> Password dialogue when opening public links - should focus be sent to the wrapper or to the input

- Not necessary to control focus when doing a reload
- Needs to provide a descriptive title

**General**
- Page titles need to be taken care of over the whole app

**Comment by marcus-herrmann** [comment]
1. The page <title> MUST be accurate and informative.
2. If a page is the result of a user action or scripted change of context, the text of the <title> SHOULD describe the result or change of context to the user.
3. The <title> SHOULD be concise.
4. The page <title> SHOULD be unique, if possible.
5. Unique information SHOULD come first in the <title>.
6. The page <title> SHOULD match (or be very similar to) the top heading in the main content.

**Comment by fschade** [comment]
all mentioned ideas are implemented in the a11y swarming in the past weeks. They are part of current master already. Closing this
